### DISCUSSION TITLE
A way to declare global functions?

### DISCUSSION THREAD

**Discussion by nam-co** [alpinejs/alpine#1123] [4. General]
Hi, I was wondering if there's a way to declare global functions that you can use in all x-data, for example vue has Vue.mixin({
  methods: {...}}), for example, Im trying to make this show hide tooltip work, but I can get it, it seems better to do it with a global function, than going inside each x-data and setting this function or variable, appreciate any help, thanks

   ```
 <div x-data="">
        <div class="tooltip" x-on:mouseover="showHide(true,'tooltipxx')" x-on:mouseleave="showHide(false,'tooltipxx')">
            <div id="tooltipxx" class="tooltip-box">the message</div>
            <button class="btn">Toggle</button>
        </div>
    </div>
<script>
        function showHide(action,element_id) {
            var element = document.getElementById(element_id);
            if(action=='toggle') {
                element.classList.toggle("visible");
                console.log('toggle '.element_id);
            } else if(action) {
                element.classList.add("visible");
                console.log('show '.element_id);
            } else {
                console.log('hide '.element_id);
                element.classList.remove("visible");
            }
        }
</script>
```

**Comment by ryangjchandler** [comment]
@nam-co You might be interested in this part of one of my blog posts: https://ryangjchandler.co.uk/articles/writing-reusable-alpine-components#mixins

**Reply by ryangjchandler** [reply]
@nam-co Although, in your situation you should be able to call the `showHide` function, as long as it's defined in the global namespace: https://codepen.io/ryangjchandler/pen/eYBGKQp?editors=1111

**Reply by nam-co** [reply]
Hi @ryangjchandler thanks for the answer, I did see your "3 Simple Alpine.js Tips and Tricks" article before, I will see this one.. but I tried the codepen, and it doesn't work on my case, if you have some time please see https://codepen.io/nam_co/pen/jOVGQdV?editors=1011

**Reply by ryangjchandler** [reply]
Seems to be working fine. The console is receiving logs. 

The problem seems to be that you're concatenation strings using `.` (like PHP) instead of `+` in JavaScript.

**Reply by nam-co** [reply]
Yes, but is not doing the "element.classList.remove("visible");" witch is should show the tooltip

**Reply by ryangjchandler** [reply]
The visible class is being applied. If you `console.log(element)`, you can see that it has the right class but your `<div>` is full width so the mouse events are a bit buggy.

**Reply by SimoTod** [reply]
It works correctly, if you right-click and inspect the page you can see that the class is added and removed when it should.
The issue is that you are defining `.tooltip .tooltip-box` with visibility hidden. When using css, rules with more selectors get precedence over rules with a single selector so `.tooltip .tooltip-box` (2 classes) is always more important than `.visible` (1 class). If you use a utility-first framework like tailwind CSS, you should avoid rules with multiple classes if possible because they are always going to overrule tailwind. Also, when 2 rules have the same number of selectors, the one defined later in the page, is always going to take precedence. Again, with tailwind, you should try to use the utility as much as possible (ad example you can toogle invisible) and define your custom rules before the framework.
See https://codepen.io/SimoTod/pen/qBqPvrx?editors=1111

**Reply by ryangjchandler** [reply]
Ha, @SimoTod beat me too it.

**Reply by nam-co** [reply]
Thank you both @SimoTod and @ryangjchandler , I didn't realize this, thanks
